src/node_blob.cc

```diff
@@ -17,21 +18,92 @@ using v8::ArrayBuffer;
 using v8::ArrayBufferView;
 using v8::BackingStore;
 using v8::Context;
-using v8::EscapableHandleScope;
 using v8::Function;
 using v8::FunctionCallbackInfo;
 using v8::FunctionTemplate;
+using v8::Global;
 using v8::HandleScope;
+using v8::Int32;
 using v8::Isolate;
 using v8::Local;
-using v8::MaybeLocal;
-using v8::Number;
 using v8::Object;
 using v8::String;
 using v8::Uint32;
 using v8::Undefined;
 using v8::Value;
 
+namespace {
+
+// Concatenate multiple ArrayBufferView/ArrayBuffers into a single ArrayBuffer.
+// This method treats all ArrayBufferView types the same.
+void Concat(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args[0]->IsArray());
+  Local<Array> array = args[0].As<Array>();
+
+  struct View {
+    std::shared_ptr<BackingStore> store;
+    size_t length;
+    size_t offset = 0;
+  };
+
+  std::vector<View> views;
+  size_t total = 0;
+
+  for (uint32_t n = 0; n < array->Length(); n++) {
+    Local<Value> val;
+    if (!array->Get(env->context(), n).ToLocal(&val)) return;
+    if (val->IsArrayBuffer()) {
+      auto ab = val.As<ArrayBuffer>();
+      views.push_back(View{ab->GetBackingStore(), ab->ByteLength(), 0});
+      total += ab->ByteLength();
+    } else {
+      CHECK(val->IsArrayBufferView());
+      auto view = val.As<ArrayBufferView>();
+      views.push_back(View{view->Buffer()->GetBackingStore(),
+                           view->ByteLength(),
+                           view->ByteOffset()});
+      total += view->ByteLength();
+    }
+  }
+
+  std::shared_ptr<BackingStore> store =
+      ArrayBuffer::NewBackingStore(env->isolate(), total);
+  uint8_t* ptr = static_cast<uint8_t*>(store->Data());
+  for (size_t n = 0; n < views.size(); n++) {
+    uint8_t* from =
+        static_cast<uint8_t*>(views[n].store->Data()) + views[n].offset;
+    std::copy(from, from + views[n].length, ptr);
+    ptr += views[n].length;
+  }
+
+  args.GetReturnValue().Set(
+      ArrayBuffer::New(env->isolate(), std::move(store)));
+}
+
+void BlobFromFilePath(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  auto entry = DataQueue::CreateFdEntry(env, args[0]);
+  if (entry == nullptr) {
+    return THROW_ERR_INVALID_ARG_VALUE(env, "Unabled to open file as blob");
+  }
+
+  std::vector<std::unique_ptr<DataQueue::Entry>> entries;
+  entries.push_back(std::move(entry));
+
+  auto blob =
+      Blob::Create(env, DataQueue::CreateIdempotent(std::move(entries)));
+
+  auto array = Array::New(env->isolate(), 2);
+  USE(array->Set(env->context(), 0, blob->object()));
+  USE(array->Set(env->context(),
+                 1,
+                 Uint32::NewFromUnsigned(env->isolate(), blob->length())));
+
+  if (blob) args.GetReturnValue().Set(array);
```

Comment:
This doesn’t look right – if `blob` could be false-y, then `blob->object()` is probably not a valid thing to call in the lines above this one.